<!--

Programmer: Victoria Albanese
Filename: camera_top.launch

Purpose: Launch the "top" Intel RealSense 435i camera
on the rigging so that its pointcloud can be visualized

-->

<launch>

    <!-- This launches the camera with the correct parameters -->
    <include file="$(find realsense2_camera)/launch/rs_rgbd.launch">
	<arg name="camera" value="camera_top" />
	<arg name="serial_no" value="841612070149" />
    </include>


    <!-- These nodelets limit the FOV of the camera & filter out the ground plane -->
    <node pkg="nodelet" type="nodelet" name="cloud_top_filter_fov_depth" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_top/depth_registered/points"/>
        <remap from="~output" to="/camera_top/depth_registered/points_filtered_depth"/>
        <rosparam>
            filter_field_name: z
            filter_limit_min: 0.0
            filter_limit_max: 0.38
            filter_limit_negative: False
        </rosparam>
    </node>	
    <node pkg="nodelet" type="nodelet" name="cloud_top_filter_fov_sides" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_top/depth_registered/points_filtered_depth"/>
        <remap from="~output" to="/camera_top/depth_registered/points_filtered_sides"/>
        <rosparam>
            filter_field_name: x
            filter_limit_min: -0.28
            filter_limit_max:  0.28
            filter_limit_negative: False
        </rosparam>
    </node>	
    <node pkg="nodelet" type="nodelet" name="cloud_top_ground_filter" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_top/depth_registered/points_filtered_sides"/>
        <remap from="~output" to="/camera_top/depth_registered/points_filtered_ground_plane"/>
        <rosparam>
            filter_field_name: y
            filter_limit_min: -0.28
            filter_limit_max:  0.28
            filter_limit_negative: False
        </rosparam>
    </node>	


    <!-- Filter out any noise with statistical outlier removal -->
    <node pkg="nodelet" type="nodelet" name="cloud_top_outlier_removal_pass1" args="load pcl/StatisticalOutlierRemoval pcl_manager" output="screen"> 
        <remap from="~input"  to="/camera_top/depth_registered/points_filtered_ground_plane" /> 
        <remap from="~output" to="/camera_top/depth_registered/points_filtered" /> 
        <rosparam> 
               mean_k: 100 
               stddev: 0.01 
               negative: False 
        </rosparam> 
    </node> 


    <!-- This transforms each resulting pointcloud into a global reference frame -->
    <node name="camera_top_tf" pkg="tf" type="static_transform_publisher"
        args="0 0 0.386 1.581 0 3.141 /world /camera_top_color_optical_frame 100"/>

</launch>

